### PR TITLE
Swap-in / Swap-out

### PR DESCRIPTION
**Pull request by nullpointer0x00**
### Summary

This PR implements the basic functionality for **swap in / swap out** operations in the vault module. It supports proportional conversions between the base asset and shares based on the current vault state.

### Details

* `SwapIn`: Converts a given amount of the base asset into a proportional amount of vault shares based on total assets and total share supply. The first deposit receives a 1:1 ratio.
* `SwapOut`: Converts a given amount of vault shares back into the base asset using proportional logic. The returned asset amount depends on the share value relative to the vault’s total assets and share supply.
* No fees, limits, or special logic are applied—this is the minimal implementation to support basic deposit/withdraw behavior.

This sets the foundation for more advanced logic in future iterations.

### CODE REVIEW

**Review by iramiller — COMMENTED** [review]
Overall seems pretty solid.  The shares should be an actual "share of" calculation and not a direct one-one amount of tokens in the vault.

File: keeper/vault.go [hunk 1]
```diff
@@ -133,3 +136,99 @@ func (k *Keeper) createVaultMarker(ctx sdk.Context, markerManager sdk.AccAddress
...
+	}
+	if vault == nil {
+		return nil, fmt.Errorf("vault with address %v not found", vaultAddr.String())
+	}
+
+	if err := vault.ValidateUnderlyingAssets(asset); err != nil {
+		return nil, err
+	}
+
+	shares := sdk.NewCoin(vault.ShareDenom, asset.Amount)
```
[the remark this answers is not in the document]
Comment (nullpointer0x00) [reply keeper/vault.go]: I added a `CalculateSharesFromAssets` method to calculate shares from assets along with tests.

File: utils/math.go [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+//
+//	assets = (shares * totalAssets) / totalShares
+func CalculateAssetsFromShares(
+	shares math.Int,
+	totalShares math.Int,
+	totalAssets math.Int,
+	assetDenom string,
+) (sdk.Coin, error) {
+	if totalShares.IsZero() {
+		return sdk.Coin{}, fmt.Errorf("cannot calculate assets: totalShares is zero")
```
Comment (iramiller) [inline utils/math.go:45]: If you have zero shares you actually have zero assets, this doesn't feel like an error condition ... now if you had less than zero shares that would certainly be an error.

File: utils/math.go [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+//
+//	shares = (assets * totalShares) / totalAssets
+func CalculateSharesFromAssets(
+	assets math.Int,
+	totalAssets math.Int,
+	totalShares math.Int,
+	shareDenom string,
+) (sdk.Coin, error) {
+	if totalAssets.IsZero() {
+		return sdk.NewCoin(shareDenom, assets), nil // First deposit: 1:1 mapping
```
Comment (iramiller) [inline utils/math.go:25]: we should verify that we don't need more precision for shares than the underlying denomination... I _suspect_ that 1:1 is fine but we should think through if something like 1:100 might make sense to increase the shares precision ... if we go that way we would need to worry about type size restrictions and overflow though.

File: utils/math_test.go [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+			expectErr:   false,
+		},
+		{
+			name:        "assets < totalAssets",
+			assets:      sdkmath.NewInt(100),
+			totalAssets: sdkmath.NewInt(1000),
+			totalShares: sdkmath.NewInt(1000),
+			expected:    sdk.NewCoin(denom, sdkmath.NewInt(100)), // (100 * 1000 / 1000)
+			expectErr:   false,
+		},
```
Comment (iramiller) [inline utils/math_test.go:98]: we should add a few more test cases for invalid or negative numbers and their expected error conditions to fully cover the range of input values our method can accept even if they are not valid.

**Review by iramiller — APPROVED** [review]
Awesome start.
